### PR TITLE
Bratseth/smarter chunking

### PR DESCRIPTION
**Pull request by bratseth**
Somewhat smarter chunking. I recommend following each commit.

### CODE REVIEW

File: linguistics/src/test/java/ai/vespa/language/chunker/FixedLengthChunkerTest.java [hunk 1]
```diff
@@ -22,23 +22,54 @@ public void testChunker() {
...
+
+        // -> First chunk is 20 long
+        assertChunks(20, "first sentence, end. Next sentence.    ",
+                     "first sentence, end.", " Next sentence.    ");
+        // -> First chunk is 21 long as we get to a soft boundary
+        assertChunks(20, "first sentence, stop.Next sentence.    ",
+                     "first sentence, stop.", "Next sentence.    ");
+        // -> First chunk is 22 long, as there is no boundary before reaching the hard limit
+        assertChunks(20, "first sentence, endNext sentence.      ",
+                     "first sentence, endNex", "t sentence.      ");
```
Comment (glebashnik) [inline linguistics/src/test/java/ai/vespa/language/chunker/FixedLengthChunkerTest.java:45]: Splitting words like this can lead to meaningless tokens and degrade retrieval quality.

File: linguistics/src/main/java/ai/vespa/language/chunker/FixedLengthChunker.java [hunk 2]
```diff
@@ -8,8 +8,18 @@
...
+ * target chunk length measured in characters (or precisely at that length, for CJK languages).
+ *
+ * If there are no double non-letter/digit characters within 5% above the target length,
+ * the chunk split will be at the first single non-letter/digit character.
+ *
+ * If there are no double non-letter/digit characters within 10% above the target length,
+ * the chunk split will be at that position, so the absolute max chunk length will be 10% above the target
+ * length.
+ *
+ * The given target chunk length is adjusted down to prefer a more even chunk length distribution to account for the
```
Comment (glebashnik) [inline linguistics/src/main/java/ai/vespa/language/chunker/FixedLengthChunker.java:21]: This would make more sense with word/token-level chunks, since we embed and index on word/token level.
I suspect that for character-level chucks, these 5% adjustments are negligible relative to variations in word length.

**Review by glebashnik — CHANGES_REQUESTED** [review]

File: linguistics/src/test/java/ai/vespa/language/chunker/FixedLengthChunkerTest.java (hunk 1 above)
Comment (bratseth) [reply linguistics/src/test/java/ai/vespa/language/chunker/FixedLengthChunkerTest.java:45]: In these tests we use unrealistically short chunks lengths to test boundary conditions.
With a chunk size of 1000 characters this condition will kick in at length 1100, so you need to have a contiguous sequence of 100 letter/digit characters before it is split like this.
The purpose of this is robustness given such degenerate input.

File: linguistics/src/main/java/ai/vespa/language/chunker/FixedLengthChunker.java (hunk 2 above)
Comment (bratseth) [reply linguistics/src/main/java/ai/vespa/language/chunker/FixedLengthChunker.java:21]: I don't quite understand what you mean. Is this to the 5% margin before we split on words not sentences, or ensuring similar distances?

Comment (bratseth) [reply linguistics/src/main/java/ai/vespa/language/chunker/FixedLengthChunker.java:21]: ... sorry, I got it now. Yes, often this won't matter much, but if your avg. size is slightly above your fixed-length it matters a lot.

Comment (glebashnik) [reply linguistics/src/main/java/ai/vespa/language/chunker/FixedLengthChunker.java:21]: Ok, got it now.
"double non-letter/digit" is a proxy of (part of) sentence separator.
"single non-letter/digit" character is a proxy for word separator.
It does make sense to have a bit longer chunk to finish sentence/word.

File: linguistics/src/test/java/ai/vespa/language/chunker/FixedLengthChunkerTest.java (hunk 1 above)
Comment (glebashnik) [reply linguistics/src/test/java/ai/vespa/language/chunker/FixedLengthChunkerTest.java:45]: Right, this is only a problem with unrealistically small chunks.

**Review by glebashnik — APPROVED** [review]
